File: strings/2645. Minimum Additions to Make Valid String.py
class Solution:
    def addMinimum(self, word: str) -> int:
        # it comes in sets
        res = 0
        # so every a must come before every b, and every c must come before every
        n = len(word)
        seen = set()
        stack = []

        ptr1 = 0
        while ptr1 < n:
            starting = word[ptr1]
            seen = set()
            seen.add(starting)
            temp = ptr1 + 1
            prev = starting
            while temp < n and word[temp] > prev:
                prev = word[temp]
                temp += 1

                # once I see a repeat, add the missing numbers
            res += 3 - (temp - ptr1)
            ptr1 = temp

        return res

File: strings/2645. Minimum Additions to Make Valid String.cpp
class Solution {
public:
    int addMinimum(string word) {
        int res = 0;
        int ptr1 = 0;
        int n = word.size();
        unordered_set<char> seen;
        while (ptr1 < n) {
            char starting = word[ptr1];
            seen.clear();
            seen.insert(starting);
            int temp = ptr1 + 1;
            char prev = starting;
            while (temp < n && word[temp] > prev) {
                prev = word[temp];
                temp++;
            }

            res += 3 - (temp - ptr1);
            ptr1 = temp;

        }


        return res;


    }
};